source/dofs/dof_handler.cc

```diff
@@ -2564,24 +2564,13 @@ DoFHandler<dim, spacedim>::distribute_dofs(
   // true cell owners temporarily. We save the current set of subdomain ids, set
   // subdomain ids to the "true" owner of each cell, and later restore these
   // flags.
-  std::vector<types::subdomain_id>                      saved_subdomain_ids;
-  const parallel::shared::Triangulation<dim, spacedim> *shared_tria =
-    (dynamic_cast<const parallel::shared::Triangulation<dim, spacedim> *>(
-      &this->get_triangulation()));
-  if (shared_tria != nullptr && shared_tria->with_artificial_cells())
-    {
-      saved_subdomain_ids.resize(shared_tria->n_active_cells());
-
-      const std::vector<types::subdomain_id> &true_subdomain_ids =
-        shared_tria->get_true_subdomain_ids_of_cells();
-
-      for (const auto &cell : shared_tria->active_cell_iterators())
-        {
-          const unsigned int index   = cell->active_cell_index();
-          saved_subdomain_ids[index] = cell->subdomain_id();
-          cell->set_subdomain_id(true_subdomain_ids[index]);
-        }
-    }
+  parallel::shared::Triangulation<dim, spacedim> *shared_tria =
+    (dynamic_cast<parallel::shared::Triangulation<dim, spacedim> *>(
+      const_cast<dealii::Triangulation<dim, spacedim> *>(
+        &this->get_triangulation())));
+  if (shared_tria != nullptr)
+    shared_tria->set_subdomain_mask(
+      parallel::shared::Triangulation<dim, spacedim>::SubdomainMask::true_ids);
```

Comment:
Personally, I would prefer:
```suggestion
  if (auto shared_tria = dynamic_cast<parallel::shared::Triangulation<dim, spacedim> *>(
      const_cast<dealii::Triangulation<dim, spacedim> *>(&this->get_triangulation()))))
    shared_tria->set_subdomain_mask(
      parallel::shared::Triangulation<dim, spacedim>::SubdomainMask::true_ids);
```